### PR TITLE
Cwalv/15265/decouple magento code and implement code to instantiate ns8 protect iframe via sdk

### PR DESCRIPTION
**Pull request by cwalv**
This mostly works, but needs additional tests (I pushed with `--no-verify`).  I wasn't able to test this very well on my lightsail instance either, but I was able to step through the new code and see the iframe resize etc.

### CODE REVIEW

File: src/client.ts [hunk 1]
```diff
@@ -22,49 +23,53 @@ class Client implements ProtectClient {
...
+    // eslint-disable-next-line consistent-return
+    const getPathForPage = (): string => {
+      // requiring a default case here breaks TS exhaustive checking
+      /* istanbul ignore next: write tests for this after `iframe-resizer` is integrated */
+      // eslint-disable-next-line default-case
+      switch (page) {
+        case ClientPage.DASHBOARD:
+          return '/';
+        case ClientPage.ORDER_DETAILS:
+          return `/order-details/${atob(orderId)}`;
```
[the remark this answers is not in the document]
Comment (crfroehlich) [reply src/client.ts]: Yes, I think naming the function parameter `platformId` would be preferable.

File: src/client.ts [hunk 2]
```diff
@@ -1,70 +1,86 @@
...
 import { protectLogger } from './logger';
 
-// KLUDGE: Postmate is going away. For now, this is a hack to support differences between
-// the way that node require vs browser require behave
-// const Postmate = require('postmate').default || require('postmate');
+/* istanbul ignore next: testing this would be basically just copying the code */
+// eslint-disable-next-line consistent-return
+const getPathForPage = (page: ClientPage, orderId: string): string => {
+  // requiring a default case here breaks TS exhaustive checking
+  // eslint-disable-next-line default-case
```
Comment (crfroehlich) [inline src/client.ts]: I think we could safely add a `default` case that returns `/` and eliminate this error. A code comment above the `default` explaining seems preferable to escaping the eslint rule?

File: src/client.ts [hunk 3]
```diff
@@ -1,70 +1,86 @@
...
+    case ClientPage.ORDER_DETAILS:
+      return `/order-details/${btoa(orderId)}`;
+    case ClientPage.ORDER_RULES:
+      return '/rules';
+    case ClientPage.SUSPICIOUS_ORDERS:
+      return '/report/suspicious-orders';
+  }
+};
+
+const validatePage = (page?: ClientPage, orderId?: string): ClientPage => {
```
Comment (crfroehlich) [inline src/client.ts]: Would like to see docs on this as well. Our future selves will thank us if we do.

File: src/internal/iframe.ts [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+/**
+ *
+ * Attaches an iframe to the page using the provided in the Client Config.
+ * {@link ClientConfig}
+ * {@link attachToId}
+ * Sets up the onMessage and onResized handlers to enable iframe-resizer to
+ * call event handlers triggered from within the NS8 protect client as well as update the iframe
+ * height when the protect client page resizes.
+ */
+
```
Comment (crfroehlich) [inline src/internal/iframe.ts]: Extra line

File: src/internal/iframe.ts [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+  /* istanbul ignore next */
+  window.iFrameResize(
+    {
+      checkOrigin: false,
+      log: debug,
+      tolerance: 5,
+      onMessage: ({ message }: { message: { name?: string; data: unknown } }) => {
+        if (message == null) {
+          // TODO: log
```
Comment (crfroehlich) [inline src/internal/iframe.ts]: Do we have access to our logger at this point? If not, I think the overhead of trying to inject our logger would be quite high. We might just want to try/catch a `console.error` if we don't have any more robust options and the comment on why it is we're doing that here?

File: src/client.ts (hunk 2 above)
Comment (cwalv) [reply src/client.ts]: If you have a `default` case, then typescript won't be able to check that we've handled all cases in the `enum`.  Written like this, if we change the `ClientPage` enum at all, typescript will fail to compile until we've handled the change here as well; adding a default case will break that.

Comment (cwalv) [reply src/client.ts]: https://www.typescriptlang.org/docs/handbook/advanced-types.html#exhaustiveness-checking
The `eslint` rules are relatively crude heuristics that try to catch the same problems, but don't serve a purpose with typescript

Comment (crfroehlich) [reply src/client.ts]: That makes sense.

File: src/internal/iframe.ts (hunk 5 above)
Comment (cwalv) [reply src/internal/iframe.ts]: we do have access to the logger; fixed the code

File: src/client.ts [hunk 6]
```diff
@@ -4,19 +4,21 @@ import { createIFrame } from './internal/iframe';
...
       return '/';
     case ClientPage.ORDER_DETAILS:
-      return `/order-details/${btoa(orderId)}`;
+      return `/order-details/${btoa(platformId)}`;
     case ClientPage.ORDER_RULES:
       return '/rules';
     case ClientPage.SUSPICIOUS_ORDERS:
       return '/report/suspicious-orders';
+    default:
+      return assertNever(page);
```
Comment (crfroehlich) [inline src/client.ts]: Nice. Would it not work to reduce line 21 to:

`throw new Error(`Invalid page "${page}"`)` 

vs calling a method to do the same?

Comment (cwalv) [reply src/client.ts]: unfortunately no .. it needs to call a function that takes `never` for typescript to do the exhaustive checking

Comment (crfroehlich) [reply src/client.ts]: Ok, makes sense.

**Pull request merged**